tools/checkstyle/checkstyle.bzl

```diff
@@ -0,0 +1,199 @@
+        'files' : 'java source files'
+    }
+)
+
+# This aspect is responsible for collecting Java sources for target
+# When applied to target, `JavaSourceFiles` struct will be attached to
+# target info
+def collect_sources_impl(target, ctx):
+    files = []
+    if hasattr(ctx.rule.attr, 'srcs'):
+        for src in ctx.rule.attr.srcs:
+            for f in src.files:
+                if f.extension == 'java':
+                    files.append(f)
+    return [JavaSourceFiles(files = files)]
+
+
+collect_sources = aspect(
+    implementation = collect_sources_impl,
+)
+
+
+# `ctx` is rule context: https://docs.bazel.build/versions/master/skylark/lib/ctx.html
+def _checkstyle_test_impl(ctx):
+    # verify target name matches naming conventions
+    if "{}-checkstyle".format(ctx.attr.target.label.name) != ctx.attr.name:
+        fail("target should follow `{java_library target name}-checkstyle` pattern")
+
+    properties = ctx.file.properties
+    suppressions = ctx.file.suppressions
+    opts = ctx.attr.opts
+    sopts = ctx.attr.string_opts
+
+    # Checkstyle and its dependencies
+    classpath = ":".join([file.path for file in ctx.files._classpath])
+
+    args = ""
+    inputs = []
+    if ctx.file.config:
+      args += " -c %s" % ctx.file.config.path
+      inputs.append(ctx.file.config)
+    if properties:
+      args += " -p %s" % properties.path
+      inputs.append(properties)
+    if suppressions:
+      inputs.append(suppressions)
+
+    # Build command to run Checkstyle test
+    cmd = " ".join(
+        ["java -cp %s com.puppycrawl.tools.checkstyle.Main" % classpath] +
+        [args] +
+        ["--%s" % x for x in opts] +
+        ["--%s %s" % (k, sopts[k]) for k in sopts] +
+        [file.path for file in ctx.attr.target[JavaSourceFiles].files]
+    )
+
+    # Wrap checkstyle command in a shell script so allow_failure is supported
+    ctx.actions.expand_template(
+        template = ctx.file._checkstyle_sh_template,
+        output = ctx.outputs.checkstyle_script,
+        substitutions = {
+            "{command}" : cmd,
+            "{allow_failure}": str(int(ctx.attr.allow_failure)),
+        },
+        is_executable = True,
+    )
+
+    files = [ctx.outputs.checkstyle_script, ctx.file.license] + ctx.attr.target[JavaSourceFiles].files + ctx.files._classpath + inputs
+    runfiles = ctx.runfiles(
+        files = files,
+        collect_data = True
+    )
+    return DefaultInfo(
+        executable = ctx.outputs.checkstyle_script,
+        files = depset(files),
+        runfiles = runfiles,
+    )
+
+checkstyle_test = rule(
+    implementation = _checkstyle_test_impl,
+    test = True,
+    attrs = {
+        "config": attr.label(
+            allow_single_file=True,
+            doc = "A checkstyle configuration file",
+            default = "//tools/checkstyle:config.xml",
+        ),
+        "suppressions": attr.label(
+            allow_single_file=True,
+            doc = "A checkstyle suppressions file"
```

Comment:
Please add:
` for specifying files and lines that should be suppressed from checks`
Also, if you have an example for such a file, please provide one.

Reply:
@oferb fixed in b1f8cbdeb78d5810a5d5827a1d198b3e65a7d851